File: test/scripts/busy-wait.py
#!/usr/bin/env python3
"""
ANTI-PATTERN: Busy-wait loop
Resume: "Avoid Busy-Wait Loops" - Continuously consumes CPU while waiting,
never yields to scheduler. Check vmstat: high %user, high context switches.

Run: python busy-wait.py
Measure: top, vmstat 1
"""

import threading
import time

DURATION_SEC = 3
NUM_WAITERS = 4  # Multiple threads busy-waiting


def busy_wait_until(deadline):
    """BAD: Spin loop - burns CPU, never yields."""
    while time.perf_counter() < deadline:
        pass  # Busy-wait: 100% CPU, no sleep


def main():
    print("ANTI-PATTERN: Busy-wait (spin loop)")
    print(f"Running {NUM_WAITERS} threads busy-waiting for {DURATION_SEC}s")
    print("Run: top -p $(pgrep -f busy-wait) -> high %CPU")

    deadline = time.perf_counter() + DURATION_SEC
    threads = [threading.Thread(target=busy_wait_until, args=(deadline,)) for _ in range(NUM_WAITERS)]

    start = time.perf_counter()
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    elapsed = time.perf_counter() - start

    print(f"Done in {elapsed:.2f}s - CPU was maxed during entire wait")
    print("Fix: use time.sleep() or threading.Event.wait()")


if __name__ == "__main__":
    main()

File: test/scripts/busy-wait.js
#!/usr/bin/env node
/**
 * ANTI-PATTERN: Busy-wait loop
 * Resume: "Avoid Busy-Wait Loops" - Continuously consumes CPU while waiting.
 *
 * Run: node busy-wait.js
 * Measure: top, vmstat 1
 */

const DURATION_MS = 3000;
const NUM_WAITERS = 4;

function busyWaitUntil(deadline) {
  while (Date.now() < deadline) {
    // Busy-wait: 100% CPU, no setTimeout/setImmediate
  }
}

console.log('ANTI-PATTERN: Busy-wait (spin loop)');
console.log(`Running ${NUM_WAITERS} "threads" (setInterval) busy-waiting for ${DURATION_MS}ms`);
console.log('Run: top -p $(pgrep -f busy-wait) -> high %CPU');

const deadline = Date.now() + DURATION_MS;
const start = Date.now();

// Simulate multiple busy-wait loops (Node is single-threaded, so one loop maxes CPU)
let completed = 0;
const check = () => {
  busyWaitUntil(deadline);
  completed++;
  if (completed < NUM_WAITERS) {
    setImmediate(check);
  } else {
    const elapsed = (Date.now() - start) / 1000;
    console.log(`Done in ${elapsed.toFixed(2)}s - CPU was maxed`);
    console.log('Fix: use setTimeout or Promise + async/await');
  }
};
check();
